task_processing/plugins/mesos/execution_framework.py

```diff
@@ -115,31 +119,79 @@ def kill_tasks_stuck_in_staging(self):
             if self.stopping:
                 return
 
-            with self._lock:
-                time_now = time.time()
-                for task_id in self.task_metadata.keys():
-                    if self.task_metadata[task_id].task_state not in (
-                        'TASK_STAGING',
-                    ):
-                        continue
-
-                    md = self.task_metadata[task_id]
-                    if time_now > (
-                        md.task_state_history['TASK_STAGING'] +
-                        self.task_staging_timeout_s
-                    ):
-                        log.warning(
-                            'Killing stuck task {id}'.format(id=task_id)
-                        )
-                        self.kill_task(task_id)
-                        self.blacklist_slave(
-                            agent_id=self.task_metadata[task_id].agent_id,
-                            timeout=self.slave_blacklist_timeout_s,
-                        )
-                        get_metric(TASK_STUCK_COUNT).count(1)
+            time_now = time.time()
+            for task_id in self.task_metadata.keys():
+                md = self.task_metadata[task_id]
+
+                eligible_for_termination_at = md.task_state_history[
+                    md.task_state] + self.task_staging_timeout_s
+
+                if time_now < eligible_for_termination_at:
+                    continue
+
+                if md.task_state == 'UNKNOWN':
+                    log.warning('Task {id} has been in unknown state for '
+                                'longer than {timeout}. Re-enqueuing it.'
+                                .format(
+                                    id=task_id,
+                                    timeout=self.task_staging_timeout_s
+                                )
+                                )
+                    self._reenqueue_task(task_id)
+                    get_metric(TASK_LAUNCH_FAILED_COUNT).count(1)
+                    continue
 
+                if md.task_state not in (
+                    'TASK_STAGING',
+                ):
+                    continue
+
+                if time_now > (
```

Comment:
Line 144-149 can be combined as (task_state==STAGING && timed out) for easier read.